### DISCUSSION TITLE
[FR] Config setting for changing the name of x-craft-preview and x-craft-live-preview query params

### DISCUSSION THREAD

**Discussion by aelvan** [craftcms/cms#15605] [Ideas] [answered]
Could we get config settings for changing the parameter names, similar to what we have with `tokenParam` and the various triggers?

Since yesterday we're seeing DDOS attacks/vulnerability probing using x-craft-preview on several of our sites. With the cache poisoning issue being taken care of via [#15589](https://github.com/craftcms/cms/pull/15589), it would be good to be able to mitigate the DDOS attacks by obfuscating the parameters, especially since they now will bypass the template caching. 

Hopefully there isn't a vulnerability related to outputting the preview parameter from untrusted sources to twig/urls, but considering that it's just a random string that isn't validated in any way, maybe it would be safer to just generate a new random string for each request, instead of using whatever is passed in? Or, add some kind of validation like the tokenParam has.

**Comment by brandonkelly** [comment, marked as answer]
> but considering that it's just a random string that isn't validated in any way, maybe it would be safer to just generate a new random string for each request, instead of using whatever is passed in?

A new random value is generated each time there is a content edit, to ensure that the browser doesn’t load the page from cache when we know something has changed. I don’t think there would be any benefit to changing it more frequently than that.

> Or, add some kind of validation like the tokenParam has.

I like that idea. Just made that change for Craft 4.12 and 5.4 (f82baa76fd08504b15412b3b7232a6dccad8d5db). So going forward, `craft\web\Request::getIsPreview()` will only return `true` if the `x-craft-preview`/`x-craft-live-preview` param is set to a valid hashed value.

**Reply by mmikkel** [reply]
> So going forward, `craft\web\Request::getIsPreview()` will only return `true` if the `x-craft-preview`/`x-craft-live-preview` param is set to a valid hashed value.

@brandonkelly Hashing the preview param is a good call, but from what I can tell it looks like the preview params are [still added](https://github.com/craftcms/cms/blob/d8c12bbb4673157f2bb7285dfbc94f6d5e722874/src/helpers/UrlHelper.php#L641) to generated URLs regardless of whether they contain valid hashes or not? 

This means that someone could hit the site with a URL like `?x-craft-preview=bogus` and potentially have Craft cache that page, with the bogus value added to all generated URLs (since `getPreview()` will now be `false`). I think this should work like the `token` param; i.e. if the preview param is not a valid hash, Craft should not add it to generated URLs at all?

**Reply by mmikkel** [reply]
FYI, the bot attacks we're seeing are specifically using bogus/trash/random values for `?x-craft-preview`. The ideal thing would be if (as @aelvan suggests) we could configure the params to be named something else, which would make it trivial to block all requests w/ the default param names.

**Reply by aelvan** [reply]
I'm not sure if this fix helps with either of the two issues I addressed. 

It's still possible to pass whatever content a malicious user wants to the preview param, and it will end up in the DOM of the page. Additionally, since `getIsPreview` no longer returns true in cases like this, the cache poisoning issue that was fixed in 4.11.5 has been reintroduced exactly for the types of scenarios that we want to avoid. 

To clarify what I meant with creating a new random value; I understand that a new random value is generated for each request, what I proposed was not to use the value that is passed through the param, when creating new URLs. Currently, if I request a page with `?x-craft-preview=vote-trump-in-november`, that string will end up in the generated markup. I don't know enough about how the url params are sanitized in Yii to know if this poses a security risk or not. But, I'm pretty sure that that text will end up in some cache somewhere, even if you go to lengths to make sure it doesn't in Craft. So what I proposed was to not use that value at all, but instead just create a new value before using it to create the new URLs. So even though the input was `?x-craft-preview=vote-trump-in-november`, the URLs in the DOM will have `?x-craft-preview=K3HJsg2Asd`.

And even though this helps slightly with DDOS attacks since the page will be cached, I would really prefer to be able to obfuscate the names to avoid it altogether.

**Reply by mmikkel** [reply]
It seems to me that as long as https://github.com/craftcms/cms/commit/f82baa76fd08504b15412b3b7232a6dccad8d5db is amended so that invalid preview param values aren't added to generated URLs, it should mitigate both issues – i.e. both the DDOS/bot attacks currently exacerbated by preview params being used to bypass caches, as well as the ability for bad actors to inject random content into URLs (and potentially having those cached and indexed, with potential for reputational damage). 

I still think configurable preview param names is a really good idea though, would love to have that in addition to the above 🙂

**Reply by brandonkelly** [reply]
> I think this should work like the `token` param; i.e. if the preview param is not a valid hash, Craft should not add it to generated URLs at all?

Yeah good call. Just made that change as well. (ca91aa3190461ccc18f9ac186778c8c9f986e50a)

**Reply by brandonkelly** [reply]
Craft 4.12.0 and 5.4.0 are out now with those `x-craft-preview` and `x-craft-live-preview` adjustments.
